common/io.hpp

```diff
@@ -115,6 +118,28 @@ class StringReader : virtual public Reader {
 	}
 };
 
+class ByteWriter : virtual public Writer {
+private:
+	vector<uint8_t> &receiver_;
+	size_t bytes_written_ {0};
+
+public:
+	ByteWriter(vector<uint8_t> &receiver) :
+		receiver_ {receiver} {
+	}
+
+	ExpectedSize Write(
+		vector<uint8_t>::const_iterator start, vector<uint8_t>::const_iterator end) override {
+		long unsigned int max_write {receiver_.size() - bytes_written_};
+		long unsigned int iterator_size {(long unsigned int) (end - start)};
+		long unsigned int bytes_to_write {std::min(iterator_size, max_write)};
```

Comment:
`size_t`, please.

Reply:
But they're not `size_t`'s

Reply:
This annoyed me to no end. I can do `size_t a = end - start` for instance. But I do get a warning it's a narrowing conversion on my setup :persevere:

Reply:
Wow, that's quite strange. Please add a comment about this.

Reply:
Yeh, so I'm guessing the reason is that the range can be negative (although we know it's not in this case, since we are passing in the iterators). But the `distance_type) (between iterators) are signed as a result, while the size is not (always >= 0)...

Pushed a fix, and added an assert for the negative range as a safe-guard. 

Please see what you think.

Reply:
According to [this page](https://stackoverflow.com/questions/22021886/subtraction-of-iterators-and-integers), the result type after subtracting two iterators is [`ptrdiff_t`](https://en.cppreference.com/w/cpp/types/ptrdiff_t). I guess it's because negative values are possible, although in practical `end - start` scenarios it should never happen. This is identical to `difference_type`, not `size_type` as you have used.

I suspect that the warnings will go away if you keep the `ptrdiff_t` type for the direct assignments (or just `auto`), but you add it to a `size_t` total with `+=`. This is allowed AFAIK.

Reply:
Hmm, I'm not sure I follow ur argument. All these things I already knew, and is what I tried to convey in the comment above.

Why should I add a signed value to unsigned (aka `ptrdiff_t` to `size_t` ?) - Isn't it better to do, as I have done, base all arithmetic in the vector size type, and then do `one`, and `one` only `explicit` cast to unsigned. Which is actually safe here, since it is guarded by the assert?

Reply:
There's also `ssize_t`.

Reply:
Yeh, but point being, I want to use `vector:size_type`

Reply:
Yeah, I don't know how to do that. But there's no guarantee it's going to be `long unsigned int` everywhere. That's exactly what `size_t` and `ssize_t` are for. They made it more flexible, but I'm not sure we can easily do the same.

Reply:
My point was that if you stick to `difference_type` or `ptrdiff_t` until you do the addition operation, you don't need any casts at all, because adding signed to unsigned is allowed, unlike assignment.

Reply:
```
using Vsize = vector<uint8_t>::const_iterator::difference_type;
```
```
error: narrowing conversion of ‘((&((mender::common::io::ByteWriter*)this)->mender::common::io::ByteWriter::receiver_)->std::vector<unsigned char>::size() - ((long unsigned int)((mender::common::io::ByteWriter*)this)->mender::common::io::ByteWriter::bytes_written_))’ from ‘long unsigned int’ to ‘mender::common::io::Vsize’ {aka ‘long int’} [-Werror=narrowing]
   55 |         Vsize max_write {receiver_.size() - bytes_written_};
```

So it is not happy with it the other way around.

Reply:
Sorry, my mistake, I did not realize that you also use `bytes_written_` at the start. Then you cannot avoid casting, I think.